import streamlit as st
import pandas as pd
import lib
import stats
import wgcna
import dynamic
import network

def home():
    st.set_page_config(layout="wide")
    st.title("Analysis of LincRNA *TUNA* Knock-down in MESCs")

    lib.initialize_data()

    st.write(
        """
        This web application contains a collection of analyses performed on the LincRNA *TUNA* knock-down dataset from \
        [Lin et al., 2014](https://doi.org/10.1016/j.molcel.2014.01.021). The dataset is comprised of RNA-seq data from \
        mouse embryonic stem cells (MESCs) with a knock-down of the LincRNA *TUNA* gene across several time-points: \
        day 2, day 4 and day 6. Control samples were taken on day 2, and these serve as the reference level for the \
        analysis.

        The analyses performed on this dataset includes a statistical analysis of differential gene expression across \
        time-points, a weighted gene co-expression network analysis (WGCNA) to identify modules of co-expressed genes, \
        and using the networks identified -- dynamic modelling of gene expression changes over time.

        The raw data for this work was taken from the [GEO database](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE46730). \
        The raw count data was generated by using a *slightly* modified version of the `nf-core` RNA-seq pipeline in the \
        [Nextflow workflow manager](https://www.nextflow.io/). All with default parameters, and against the GRCm39 mouse genome.

        ### Synopsis
        The statistical method of differential gene expression was not adequate to identify key driver genes in the *TUNA* \
        knock-down dataset. Therefore, a network-based approach was taken to identify modules of co-expressed genes. \
        The WGCNA analysis identified 3 modules of (highly correlated) co-expressed genes, with the `maroon` module being \
        the most significant. The dynamic modelling of gene expression changes over time was performed using the \
        network identified in the `maroon` module.

        To extend this work, it would be interesting to now look at gene ontology enrichment of the WGCNA modules, \
        and see if there is any obvious system changes in the *TUNA* knock-down MESCs, and build that logic into the \
        dynamic modelling (i.e. use a more complex model with known protein-protein interactions, for example).
        """
    )



pg = st.navigation([
    st.Page(home, title="Home", icon="üè†"),
    st.Page(stats.stats_page, title="Statistical Modelling", icon="üìä"),
    st.Page(wgcna.wgcna_page, title="WGCNA", icon="üìä"),
    st.Page(network.network_page, title="Network Analysis", icon="üìä"),
    st.Page(dynamic.dynamic_page, title="Dynamic Modelling", icon="üìä")
])
pg.run()
